### ISSUE TITLE
Issues with registering key using pamu2fcfg

### ISSUE THREAD

**Issue by jwels** [Yubico/pam-u2f#172]
When trying to register a second key I got the following error:
`fido_dev_make_cred_rx: fido_rx `
`error: fido_dev_make_cred (-2) FIDO_ERR_RX`
I deleted the already registered key and tried re-registering it, but ended up with the same error.
When using the debug flag I additionally get:
`rx: payload_len=1421`
`rx: count < payload_len`
I also tried reinstalling the module, via apt as well as from source and resetting the key.
Whats really interesting to me is that the first key worked and the error only showed up later on when I tried to add the second one. It looks like it is a Feitian key, this is what lsusb shows:
`Bus 001 Device 020: ID 096e:0854 Feitian Technologies, Inc. FIDO KB`
Udev rules are active and both keys work fine when used in a browser.
Thanks for any help.

**Comment by a-dma** [comment]
Do I understand correctly that you have two (identical?) devices and now neither of those would register using `pamu2fcfg`?

Could you post a longer debug dump of the problematic case please?

**Comment by jwels** [comment]
Thanks for the reply. Yes that's exactly the case.

This is the whole debug output:
```
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
get_report_descriptor: open
fido_tx: d=0x55d1cf5e96d0, cmd=0x06, buf=0x55d1cf5e96d0, count=8
0000: 0d 21 bc d3 20 b5 b2 27
fido_rx: d=0x55d1cf5e96d0, cmd=0x06, buf=0x55d1cf5e96d8, count=17, ms=-1
rx_preamble: initiation frame at 0x7ffc5aa78da0
0000: ff ff ff ff 86 00 11 0d 21 bc d3 20 b5 b2 27 00
0016: 00 00 02 02 01 00 01 07 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx: payload_len=17
fido_rx: buf=0x55d1cf5e96d8, len=17
0000: 0d 21 bc d3 20 b5 b2 27 00 00 00 02 02 01 00 01
0016: 07
fido_dev_get_cbor_info_tx: dev=0x55d1cf5e96d0
fido_tx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x7ffc5aa78e07, count=1
0000: 04
fido_dev_get_cbor_info_rx: dev=0x55d1cf5e96d0, ci=0x55d1cf5e9610, ms=-1
fido_rx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x7ffc5aa78910, count=1200, ms=-1
rx_preamble: initiation frame at 0x7ffc5aa78880
0000: 00 00 00 02 90 00 89 00 aa 01 82 66 55 32 46 5f
0016: 56 32 68 46 49 44 4f 5f 32 5f 30 02 82 6b 63 72
0032: 65 64 50 72 6f 74 65 63 74 6b 68 6d 61 63 2d 73
0048: 65 63 72 65 74 03 50 83 3b 72 1a ff 5f 4d 00 bb
rx: payload_len=137
rx: continuation frame at 0x7ffc5aa78880
0000: 00 00 00 02 00 2e bd da 3e c0 1e 29 04 a5 62 72
0016: 6b f5 62 75 70 f5 62 75 76 f4 64 70 6c 61 74 f4
0032: 69 63 6c 69 65 6e 74 50 69 6e f5 05 19 08 00 06
0048: 81 01 07 0a 08 18 60 09 81 63 75 73 62 0a 81 a2
rx: continuation frame at 0x7ffc5aa78880
0000: 00 00 00 02 01 63 61 6c 67 26 64 74 79 70 65 6a
0016: 70 75 62 6c 69 63 2d 6b 65 79 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
fido_rx: buf=0x7ffc5aa78910, len=137
0000: 00 aa 01 82 66 55 32 46 5f 56 32 68 46 49 44 4f
0016: 5f 32 5f 30 02 82 6b 63 72 65 64 50 72 6f 74 65
0032: 63 74 6b 68 6d 61 63 2d 73 65 63 72 65 74 03 50
0048: 83 3b 72 1a ff 5f 4d 00 bb 2e bd da 3e c0 1e 29
0064: 04 a5 62 72 6b f5 62 75 70 f5 62 75 76 f4 64 70
0080: 6c 61 74 f4 69 63 6c 69 65 6e 74 50 69 6e f5 05
0096: 19 08 00 06 81 01 07 0a 08 18 60 09 81 63 75 73
0112: 62 0a 81 a2 63 61 6c 67 26 64 74 79 70 65 6a 70
0128: 75 62 6c 69 63 2d 6b 65 79
parse_reply_element: cbor type
parse_reply_element: cbor type
parse_reply_element: cbor type
parse_reply_element: cbor type
fido_dev_open_rx: FIDO_MAXMSG=1200, maxmsgsiz=2048
fido_tx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x55d1cf5ecf20, count=163
0000: 01 a4 01 58 20 7f 2d 21 a1 47 35 09 80 d9 8c cc
0016: d9 41 ee c0 9a b3 be c0 f7 9e 4a 02 89 b7 2b 3e
0032: 59 3a 4d fa 27 02 a2 62 69 64 6c 70 61 6d 3a 2f
0048: 2f 70 6f 70 2d 6f 73 64 6e 61 6d 65 6c 70 61 6d
0064: 3a 2f 2f 70 6f 70 2d 6f 73 03 a3 62 69 64 58 20
0080: 79 43 54 0a de 29 6c 70 a1 d7 3d 25 24 83 08 48
0096: dc f4 79 31 55 de 61 7e 89 b9 5d 36 0e 94 23 6b
0112: 64 6e 61 6d 65 64 6a 65 6e 73 6b 64 69 73 70 6c
0128: 61 79 4e 61 6d 65 64 6a 65 6e 73 04 81 a2 63 61
0144: 6c 67 26 64 74 79 70 65 6a 70 75 62 6c 69 63 2d
0160: 6b 65 79
fido_rx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x7ffc5aa788d0, count=1200, ms=-1
rx_preamble: initiation frame at 0x7ffc5aa78840
0000: 00 00 00 02 90 00 01 36 00 00 00 00 00 00 00 00
0016: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx: payload_len=1
fido_rx: buf=0x7ffc5aa788d0, len=1
0000: 36
cbor_parse_reply: blob[0]=0x36
fido_dev_make_cred_rx: parse_makecred_reply
Enter PIN for /dev/hidraw4: 
fido_dev_authkey_tx: dev=0x55d1cf5e96d0
fido_tx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x55d1cf5fc440, count=6
0000: 06 a2 01 01 02 02
fido_dev_authkey_rx: dev=0x55d1cf5e96d0, authkey=0x55d1cf5ebb10, ms=-1
fido_rx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x7ffc5aa78800, count=1200, ms=-1
rx_preamble: initiation frame at 0x7ffc5aa78770
0000: 00 00 00 02 90 00 51 00 a1 01 a5 01 02 03 38 18
0016: 20 01 21 58 20 27 2b 78 04 12 e3 45 ec c1 13 38
0032: 92 8e e9 d4 1f 5c 0f e5 50 17 41 14 32 ad 9f fd
0048: 7a 43 dd ff 57 22 58 20 57 e7 4e 66 89 45 67 fe
rx: payload_len=81
rx: continuation frame at 0x7ffc5aa78770
0000: 00 00 00 02 00 8d 08 49 0a 04 40 2f 3a 9f 51 54
0016: cd b2 a8 21 84 69 8d f3 a5 78 bb 1f 2d 00 00 00
0032: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
fido_rx: buf=0x7ffc5aa78800, len=81
0000: 00 a1 01 a5 01 02 03 38 18 20 01 21 58 20 27 2b
0016: 78 04 12 e3 45 ec c1 13 38 92 8e e9 d4 1f 5c 0f
0032: e5 50 17 41 14 32 ad 9f fd 7a 43 dd ff 57 22 58
0048: 20 57 e7 4e 66 89 45 67 fe 8d 08 49 0a 04 40 2f
0064: 3a 9f 51 54 cd b2 a8 21 84 69 8d f3 a5 78 bb 1f
0080: 2d
fido_tx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x55d1cf5fe050, count=102
0000: 06 a4 01 01 02 05 03 a5 01 02 03 26 20 01 21 58
0016: 20 20 b9 5b 8c 56 ae e8 37 fa 5b b9 a7 ad 96 7d
0032: 55 7c 23 aa b0 77 96 b3 13 ac d3 2c 80 bf 40 83
0048: 93 22 58 20 91 98 c6 73 c3 49 f5 96 72 8b fd 36
0064: 2a c4 6e f0 52 bd 0c 45 02 ae 3a 73 72 06 1f d3
0080: a1 3c 1c a9 06 50 c5 e0 fb 0f 31 e5 ce ff 53 b6
0096: e9 ce 78 1e cd 1b
fido_rx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x7ffc5aa787a0, count=1200, ms=-1
rx_preamble: initiation frame at 0x7ffc5aa78700
0000: 00 00 00 02 90 00 25 00 a1 02 58 20 39 c1 6f db
0016: 59 ec fe 45 5f 7c df 49 70 e6 8e 8b b9 52 78 8a
0032: 76 81 56 6c af b6 9f 8c a1 e4 09 c9 00 00 00 00
0048: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
rx: payload_len=37
fido_rx: buf=0x7ffc5aa787a0, len=37
0000: 00 a1 02 58 20 39 c1 6f db 59 ec fe 45 5f 7c df
0016: 49 70 e6 8e 8b b9 52 78 8a 76 81 56 6c af b6 9f
0032: 8c a1 e4 09 c9
fido_tx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x55d1cf5fd100, count=183
0000: 01 a6 01 58 20 7f 2d 21 a1 47 35 09 80 d9 8c cc
0016: d9 41 ee c0 9a b3 be c0 f7 9e 4a 02 89 b7 2b 3e
0032: 59 3a 4d fa 27 02 a2 62 69 64 6c 70 61 6d 3a 2f
0048: 2f 70 6f 70 2d 6f 73 64 6e 61 6d 65 6c 70 61 6d
0064: 3a 2f 2f 70 6f 70 2d 6f 73 03 a3 62 69 64 58 20
0080: 79 43 54 0a de 29 6c 70 a1 d7 3d 25 24 83 08 48
0096: dc f4 79 31 55 de 61 7e 89 b9 5d 36 0e 94 23 6b
0112: 64 6e 61 6d 65 64 6a 65 6e 73 6b 64 69 73 70 6c
0128: 61 79 4e 61 6d 65 64 6a 65 6e 73 04 81 a2 63 61
0144: 6c 67 26 64 74 79 70 65 6a 70 75 62 6c 69 63 2d
0160: 6b 65 79 08 50 64 11 88 8c 84 56 74 a7 65 09 d9
0176: a9 16 c5 21 4b 09 01
fido_rx: d=0x55d1cf5e96d0, cmd=0x10, buf=0x7ffc5aa788d0, count=1200, ms=-1
rx_preamble: initiation frame at 0x7ffc5aa78840
0000: 00 00 00 02 90 05 8c 00 a3 01 66 70 61 63 6b 65
0016: 64 02 58 e4 d3 aa 72 d0 9d 92 73 9d 2b 1d 5f 09
0032: 3c 49 bc 5c 82 1b 7c 75 07 b7 6c 0a 9e 9d 61 d3
0048: 5a ea 0f 4b 45 00 00 1a 10 83 3b 72 1a ff 5f 4d
rx: payload_len=1420
rx: count < payload_len
fido_dev_make_cred_rx: fido_rx
error: fido_dev_make_cred (-2) FIDO_ERR_RX
```

**Comment by a-dma** [comment]
Thank you.

This might be a bit of a wild guess, but what OS are you using? Is it "Pop! OS"? The exact version is not hugely important, but I'm just trying to figure out if I'm decoding this correctly.

The reason I'm asking is because I see mangled strings like `pam:/\x00H/pop-` in the CBOR data. Are you using some "unusual" combinations of terminal, locale, character set, encoding or similar? Could you please paste the output of the commands `locale` and `locale charmap`. Other tests would include: a different terminal, a different locale or perhaps different `-i` and `-o` options to `pamu2fcfg`, making sure there's only ASCII characters there.

**Comment by jwels** [comment]
Your guess is correct, it is Pop OS!
I am using the standard gnome terminal and my hostname and username dont contain any special characters.  Trying different -i and -o flags result in the same error. My locale is as follows:
```
LANG=de_DE.UTF-8
LANGUAGE=de:en
LC_CTYPE="de_DE.UTF-8"
LC_NUMERIC=de_DE.UTF-8
LC_TIME=de_DE.UTF-8
LC_COLLATE="de_DE.UTF-8"
LC_MONETARY=de_DE.UTF-8
LC_MESSAGES="de_DE.UTF-8"
LC_PAPER=de_DE.UTF-8
LC_NAME=de_DE.UTF-8
LC_ADDRESS=de_DE.UTF-8
LC_TELEPHONE=de_DE.UTF-8
LC_MEASUREMENT=de_DE.UTF-8
LC_IDENTIFICATION=de_DE.UTF-8
LC_ALL=
```
locale charmap:
```UTF-8```

**Comment by a-dma** [comment]
Sorry, that was a red herring. All looks good there, my decoder was trying to be more helpful than I anticipated and happily ate some garbage. I'll keep looking into this.

**Comment by a-dma** [comment]
So, this seems to be related to https://github.com/Yubico/libfido2/issues/171 where the internal buffer used by libfido2 is 1,200 bytes long and therefore insufficient to accommodate the response from your device (thanks to @LDVG for the pointer).

This was fixed in libfido2 version 1.5.0 and later. I don't know what version of Pop! OS you're running, but if your distro is based on Focal (20.04) then there is an updated version of libfido2 on Yubico's PPA that perhaps you can use. At this time we do not have an updated version for Groovy (20.10) on the PPA, but we can try to help with that if needed. Otherwise trying to download the .deb or manually building libfido2 should also work.

**Comment by jwels** [comment]
Yes, this worked! I build the latest libfido2 version from source and now I can add both keys without problems. Thank you so much for your help!
